bpf/bpf_lxc.c

```diff
@@ -679,6 +679,30 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
 		}
 	}
 
+#ifdef ENABLE_EGRESS_GATEWAY
+	{
+		struct egress_info *info;
+		struct endpoint_key key = {};
+
+		info = lookup_ip4_egress_endpoint(ip4->saddr, ip4->daddr);
+		if (!info)
+			goto skip_egress_gateway;
+
+		/* Encap and redirect the packet to egress gateway node through a tunnel.
+		 * Even if the tunnel endpoint is on the same host, follow the same data
+		 * path to be consistent. In future, it can be optimized by directly
+		 * direct to external interface.
+		 */
+		ret = encap_and_redirect_lxc(ctx, info->tunnel_endpoint, encrypt_key,
```

Comment:
When the gateway node is the local node, this code causes us to send the packet to cilium_vxlan to then come back to the stack and finally leave through the native device. This is of course unnecessary and likely to cause issues later on.

@anfernee Is this something you want to fix in this PR? If not, we should track it somewhere. I think it's fine if we fix after the feature freeze.

Reply:
Sure. We can treat it as an optimization. I've created #15426. The problem here is it's hard to tell it's tunneling to itself from here, because we are not using the `node_ip` as `tunnel_endpoint` here. It's the `egress_ip`. It doesn't know it owns the egress IP.